scripts-dev/generate_full_schema.sh

```diff
@@ -0,0 +1,186 @@
+#!/bin/bash
+#
+# This script generates SQL files for creating a brand new Synapse DB with the latest
+# schema, on both SQLite3 and Postgres.
+#
+# It does so by having Synapse generate an up-to-date SQLite DB, then running
+# synapse_port_db to convert it to Postgres. It then dumps the contents of both.
+
+POSTGRES_HOST="localhost"
+POSTGRES_DB_NAME="synapse_full_schema"
+SQLITE_DB_FILE_NAME="full_schema_sqlite.db"
+
+SQLITE_FULL_SCHEMA_OUTPUT_FILE="full.sql.sqlite"
+POSTGRES_FULL_SCHEMA_OUTPUT_FILE="full.sql.postgres"
+OUTPUT_DIR=$(pwd)
+
+usage() {
+  echo "Usage: $0 -p <postgres_username> [-c] [-v] [-o] [-h]"
+  echo
+  echo "-p <postgres_username>"
+  echo "  Username to connect to local postgres instance. The password will be requested"
+  echo "  during script execution."
+  echo "-c"
+  echo "  Whether to enable coverage tracking. Useful for CI runs."
+  echo "-v"
+  echo "  Suppress warning about requiring the use of a virtualenv."
+  echo "-o"
+  echo "  Directory to output full schema files to. Defaults to the current directory."
+  echo "-h"
+  echo "  Display this help text."
+}
+
+while getopts "p:cvo:h" opt; do
+  case $opt in
+    p)
+      POSTGRES_USERNAME=$OPTARG
+      ;;
+    c)
+      COVERAGE=1
+      ;;
+    v)
+      NO_VIRTUALENV=1
+      ;;
+    o)
+      command -v realpath > /dev/null || (echo "The -o flag requires the 'realpath' binary to be installed" && exit 1)
+      OUTPUT_DIR="$(realpath "$OPTARG")"
+      ;;
+    h)
+      usage
+      exit
+      ;;
+    \?)
+      echo "Invalid option: -$OPTARG" >&2
+      usage
+      exit
+      ;;
+  esac
+done
+
+# Check that the script is running with a virtualenv enabled
+if [ -z ${NO_VIRTUALENV+x} ] && [ -z ${VIRTUAL_ENV+x} ]; then
```

Comment:
I got that trick from a SO answer a while back, and have used it to success ever since. Testing `-z "$foo"` seems to work just as well though, so I'll switch everything over to that.